spec/integrations/connection_edge_case_spec.rb

```diff
@@ -0,0 +1,467 @@
+    end
+
+    let(:fake_socket) do
+      instance_double(UDPSocket,
+        connect: true,
+        send: true)
+    end
+
+    let(:fake_socket_retry) do
+      instance_double(UDPSocket,
+        connect: true,
+        send: true)
+    end
+
+    context 'when having unknown SocketError (drop strategy)'do
+      before do
+        allow(fake_socket)
+          .to receive(:send)
+          .and_raise(SocketError, 'unknown error')
+      end
+
+      it 'sends using the first socket' do
+        expect(fake_socket)
+          .to receive(:send)
+          .with('foobar', anything)
+
+        subject.write('foobar')
+      end
+
+      it 'ignores the writing failure (message dropped)' do
+        expect do
+          subject.write('foobar')
+        end.not_to raise_error
+      end
+
+      it 'does not retry to send message' do
+        expect(fake_socket_retry)
+          .not_to receive(:send)
+
+        subject.write('foobar')
+      end
+
+      it 'logs the error message' do
+        subject.write('foobar')
+
+        expect(log.string).to match 'Statsd: SocketError unknown error'
+      end
+    end
+
+    context 'when having closed sockets (retry strategy)'do
+      before do
+        allow(fake_socket)
+          .to receive(:send)
+          .and_raise(IOError.new('closed stream'))
+      end
+
+      context 'when retrying is working' do
+        it 'tries with the initial socket' do
+          expect(fake_socket)
+            .to receive(:send)
+            .with('foobar', anything)
+
+          subject.write('foobar')
+        end
+
+        it 'retries on the second opened socket' do
+          expect(fake_socket_retry)
+            .to receive(:send)
+            .with('foobar', anything)
+
+          subject.write('foobar')
+        end
+      end
+
+      context 'when retrying fails' do
+        context 'because of an unknown error' do
+          before do
+            allow(fake_socket_retry)
+              .to receive(:send)
+              .and_raise(RuntimeError, 'yolo')
+          end
+
+          it 'ignores the connection failure' do
+            expect do
+              subject.write('foobar')
+            end.not_to raise_error
+          end
+
+          it 'logs the error message' do
+            subject.write('foobar')
+            expect(log.string).to match 'Statsd: RuntimeError yolo'
+          end
+        end
+
+        context 'because of a SocketError' do
+          before do
+            allow(fake_socket_retry)
+              .to receive(:send)
+              .and_raise(SocketError, 'yolo')
+          end
+
+          it 'ignores the connection failure' do
+            expect do
+              subject.write('foobar')
+            end.not_to raise_error
+          end
+
+          it 'logs the error message' do
+            subject.write('foobar')
+            expect(log.string).to match 'Statsd: SocketError yolo'
+          end
+        end
+      end
```

Comment:
Aren't they testing the same thing? I'm may miss a subtlety here.

Reply:
It is, just to be sure that we don't handle socketerrors differently.